### PR TITLE
Transcoding intermediate container independent of file extension

### PR DESCRIPTION
**Pull request by cameel**
This pull request makes the split and extract commands detect the demuxer used by ffmpeg for the input file and use that information to explicitly select the output container format. The container type no longer depends on input file extension and files with missing or wrong extensions are now handled properly.

### Details
1) The split command used to create `split-results.json` that's supposed to contain the input file metadata. When using extract+split though, split's input is the video-only file produced by the extract command and not the original input file. Its metadata is incomplete (it has no audio, data or subtitles) and should not be used for validation. To fix this problem each command now produces its own file. The transcoding task now reads metadata from `extract-and-split-results.json` because that's the file produced by the command it explicitly calls.

### Dependencies
This pull request is based on #4315 and should be merged after it.

For the code to work you also need a `golemfactory/ffmpeg-experimental` image built with https://github.com/golemfactory/ffmpeg-tools/pull/7. Tests in the CI won't pass until it's released on DockerHub.

**NOTE**: I have set the base branch of this pull request to `transcoding-muxers-and-demuxers` (#4315) rather than `CGI/transcoding/master`. Please remember to change the base branch back to `CGI/transcoding/master` before you merge.

### CODE REVIEW

**Comment by nieznanysprawiciel** [comment]
I tested this PR and I have some questions.

Conversions:
mp4 h264 -> 3gp h264
mp4 h264 -> 3g2 h264

Automatic encoder selection failed for output stream #0:1. Default encoder for format 3gp (codec amr_nb) is probably disabled. Please choose an encoder manually.
This is related to audio codec.
This fix, that you proposed, that we must check default audio codec, if it is supported - will this fix solve this problem, or is it something different?

mp4 h264 -> avi h265
Generated output has codec rawvideo and I'm not able to watch video. Probably avi doesn't support h265 and we have to block this conversion or is this bug?

**Comment by cameel** [comment]
### tl;dr
1. ffmpeg in the image (version 3.4.6) was built without the `--enable-libopencore_amrnb` flag. Getting one with this flag enabled will solve the 3gp/3g2 problem.
2. Looks like ffmpeg does not write the proper codec tag when muxing HEVC into an AVI container. The workaround is to set the tag explicitly using `-vtag hev1` option but this requires enabling experimental features (`-strict -2`). My recommendation is to mark HEVC as unsupported in the AVI container.

### Details
#### `h264/mp4` -> `h264/3gp` or `h264/3g2`
> Automatic encoder selection failed for output stream #0:1. Default encoder for format 3gp (codec amr_nb) is probably disabled. Please choose an encoder manually.
> This is related to audio codec.
> This fix, that you proposed, that we must check default audio codec, if it is supported - will this fix solve this problem, or is it something different?

I tried to reproduce it on my system and it works fine when I just convert a file from command line. But it indeed fails when I test it by adding this codec/container combination to `test_ffmpegintegration.py`. Turns out that the `ffmpeg-experimental` image has a different ffmpeg version. I have version `n4.1.3` while the image has `3.4.6-0ubuntu0.18.04.1`. Also, mine was built with different options (in particular with the `--enable-libopencore_amrnb` flag):

Image:
```
ffmpeg version 3.4.6-0ubuntu0.18.04.1 Copyright (c) 2000-2019 the FFmpeg developers
built with gcc 7 (Ubuntu 7.3.0-16ubuntu3)
configuration: --prefix=/usr --extra-version=0ubuntu0.18.04.1 --toolchain=hardened --libdir=/usr/lib/x86_64-linux-gnu --incdir=/usr/include/x86_64-linux-gnu --enable-gpl --disable-stripping --enable-avresample --enable-avisynth --enable-gnutls --enable-ladspa --enable-libass --enable-libbluray --enable-libbs2b --enable-libcaca --enable-libcdio --enable-libflite --enable-libfontconfig --enable-libfreetype --enable-libfribidi --enable-libgme --enable-libgsm --enable-libmp3lame --enable-libmysofa --enable-libopenjpeg --enable-libopenmpt --enable-libopus --enable-libpulse --enable-librubberband --enable-librsvg --enable-libshine --enable-libsnappy --enable-libsoxr --enable-libspeex --enable-libssh --enable-libtheora --enable-libtwolame --enable-libvorbis --enable-libvpx --enable-libwavpack --enable-libwebp --enable-libx265 --enable-libxml2 --enable-libxvid --enable-libzmq --enable-libzvbi --enable-omx --enable-openal --enable-opengl --enable-sdl2 --enable-libdc1394 --enable-libdrm --enable-libiec61883 --enable-chromaprint --enable-frei0r --enable-libopencv --enable-libx264 --enable-shared
libavutil      55. 78.100 / 55. 78.100
libavcodec     57.107.100 / 57.107.100
libavformat    57. 83.100 / 57. 83.100
libavdevice    57. 10.100 / 57. 10.100
libavfilter     6.107.100 /  6.107.100
libavresample   3.  7.  0 /  3.  7.  0
libswscale      4.  8.100 /  4.  8.100
libswresample   2.  9.100 /  2.  9.100
libpostproc    54.  7.100 / 54.  7.100
```

My system:
```
ffmpeg version n4.1.3 Copyright (c) 2000-2019 the FFmpeg developers
built with gcc 8.2.1 (GCC) 20181127
configuration: --prefix=/usr --disable-debug --disable-static --disable-stripping --enable-fontconfig --enable-gmp --enable-gnutls --enable-gpl --enable-ladspa --enable-libaom --enable-libass --enable-libbluray --enable-libdrm --enable-libfreetype --enable-libfribidi --enable-libgsm --enable-libiec61883 --enable-libjack --enable-libmodplug --enable-libmp3lame --enable-libopencore_amrnb --enable-libopencore_amrwb --enable-libopenjpeg --enable-libopus --enable-libpulse --enable-libsoxr --enable-libspeex --enable-libssh --enable-libtheora --enable-libv4l2 --enable-libvidstab --enable-libvorbis --enable-libvpx --enable-libwebp --enable-libx264 --enable-libx265 --enable-libxcb --enable-libxml2 --enable-libxvid --enable-nvdec --enable-nvenc --enable-omx --enable-shared --enable-version3
libavutil      56. 22.100 / 56. 22.100
libavcodec     58. 35.100 / 58. 35.100
libavformat    58. 20.100 / 58. 20.100
libavdevice    58.  5.100 / 58.  5.100
libavfilter     7. 40.101 /  7. 40.101
libswscale      5.  3.100 /  5.  3.100
libswresample   3.  3.100 /  3.  3.100
libpostproc    55.  3.100 / 55.  3.100
```

I'm not 100% sure but I think it used to be 4.1.1 in the image before. That's what we got as a requirement from you in the first spec of the transcoding task. The regression might have been caused by the change from `golemfactory/base:1.4` to `golemfactory/base:1.5` that happened in e98f31e27c3965a8de427c6b0b56bc0bfbcfa924 and probably got into `CGI/transcoding/master` when you merged `develop` into it. 1.4 was based on Debian Jessie while 1.5 is based on Ubuntu 18.04 (bionic). ffmpeg 4 was only introduced in Ubuntu 18.10 (see [ffmpeg package in Ubuntu](https://launchpad.net/ubuntu/+source/ffmpeg)). I'm not completely sure about Debian (see [ffmpeg package in Debian](https://tracker.debian.org/pkg/ffmpeg)) - it says that it's still ffmpeg 3 in stable but maybe you were getting it from testing?

In any case, you need to install newer ffmpeg in the image. People on the net seem to be recommending this untrusted PPA: [jonathonf](https://launchpad.net/~jonathonf/+archive/ubuntu/ffmpeg-4). But I'm not even sure that it has the `--enable-libopencore_amrnb` flag. Maybe it would be better to build ffmpeg in Dockerfile from source with the options you need.

#### `h264/mp4` -> `hevc/avi`
> Generated output has codec rawvideo and I'm not able to watch video. Probably avi doesn't support h265 and we have to block this conversion or is this bug?

Seems to be just how ffmpeg works right now. HEVC implementation seems to be still a bit experimental in some respects. From what I gather from information on the net (see [How to restore unplayable H265 avi files produced by ffmpeg?](https://superuser.com/questions/1110042/how-to-restore-unplayable-h265-avi-files-produced-by-ffmpeg)), for AVI it writes a proper HEVC video stream but simply does not set the codec tag and both ffmpeg and players don't recognize it as HEVC. I'm not sure why that's the case - sounds like something trivial to fix on the ffmpeg part but I have not found any official info. The general problem with AVI is that the base format is very easy to extend and was extended over time but those extensions are usually not standardized in any way. So perhaps it's just not possible to do in a standard/portable way?

I tried the fix from the link above (addding `-vtag hev1` before the output file name in the ffmpeg command) and it works both on the older ffmpeg in the image and on 4.1.3 in my system. In both cases it's necessary to enable experimental features with `-strict -2` - without it the `-vtag` option gets silently ignored and the stream ends up as `rawvideo`. The file created with `-vtag hev1 -strict -2` plays correctly in mplayer. Unfortunately it does not in VLC. The video is frozen and I'm getting lots of errors like this on the console:
```
[hevc @ 0x7fbb8cda5d40] Could not find ref with POC 25
```
```
[00007fbb8cc08070] avcodec decoder error: more than 5 seconds of late video -> dropping frame (computer too slow ?)
```

I'm not getting them when using VLC to play the original video or one converted to `hevc/mp4`.

My personal opinion is that it would be best to mark HEVC as not supported in the AVI container. It seems to be experimental at best right now.

If we want to support it, perhaps add a flag for enabling experimental features in the task definition an in the UI clearly explain to the user that he's using an experimental feature and good results are not guaranteed. In that case it would also be best to find out if it's something someone is actively working on and if there's any chance it's going to become stable.

**Comment by nieznanysprawiciel** [comment]
Merged in PR:
https://github.com/golemfactory/golem/pull/4407
